src/collapse/perf.rs

```diff
@@ -324,6 +387,91 @@ impl PerfState {
         }
         Ok(())
     }
+
+    // Use addr2line to determine the symbol to use for each program counter address
+    // (as opposed to using the symbol names that perf script produces)
+    fn un_inline(&mut self, addr: &str, module: &str) -> result::Result<String, Error> {
+        let addr = u64::from_str_radix(addr, 16).map_err(|e| {
+            Error::Parse(format!(
+                "unable to parse {} as hex address, caused by: {}",
+                addr, e
+            ))
+        })?;
+
+        if let Entry::Vacant(entry) = self.addr2line_contexts.entry(module.into()) {
+            let bytes = fs::read(module)?;
+            let file = &object::File::parse(&bytes).map_err(|e| {
+                Error::Parse(format!(
+                    "unable to parse file: {}, caused by: {}",
+                    module, e
+                ))
+            })?;
+            entry.insert(addr2line::Context::new(file)?);
+        };
+        let ctx = &self.addr2line_contexts[module];
+
+        let mut frames = ctx.find_frames(addr).map_err(|e| {
+            Error::Parse(format!(
+                "unable to parse frames from module: {}, at address: {}. caused by: {}",
+                module, addr, e
+            ))
+        })?;
+
+        let mut full_func = SmallVec::<[String; 1]>::new();
+        let mut total_length = 0;
+        while let Some(frame) = frames.next().unwrap() {
+            let mut function = frame
+                .function
+                .map(|func| {
+                    let name = remove_discriminator(
+                        func.raw_name()
+                            .map(Cow::from)
+                            .unwrap_or_else(|_| Cow::from("??")),
+                    );
+                    strip_rust_hash(addr2line::demangle_auto(name, func.language).as_ref())
+                        .to_string()
+                })
+                .unwrap_or_else(|| "??".into());
+
+            if self.opt.show_context {
+                let loc = frame.location;
+                let (file, line) = match loc {
+                    Some(ref loc) => {
+                        let file = loc
+                            .file
+                            .as_ref()
+                            .map(|f| Path::new(f).file_name())
+                            .map(|f| f.map(OsStr::to_str))
+                            .unwrap_or(None)
+                            .unwrap_or(None);
```

Comment:
Why all this song and dance to turn `loc.file` into a `Path`?

Reply:
I hate it too. First of all loc.file is an Option, so I have to deal with the None case. Then I want to get the file name without the full path so I do Path:::new(f).file_name(). But this returns Option<&OsStr>. And OsStr can't be formatted, so I call to_str, which also returns an Option since it might not be valid unicode. What's the simplest portable way to strip the path and leave only the file name?

Reply:
Hmm.. How about:
```rust
let fname = f.and_then(|f| f.rsplitn(2, std::path::MAIN_SEPARATOR).next());
```

Reply:
I like that much more, but I wish there was something simpler to use in the standard library.

Reply:
Maybe file an issue? Or maybe even a PR!